### ISSUE TITLE
workbox-window starts listening for message events too late

### ISSUE THREAD

**Issue by philipwalton** [GoogleChrome/workbox#2202]
**Library Affected**:
*workbox-window*

At the moment, the `Workbox` class adds its `message` event listener only after `register()` is called, but since the `register()` method is generally deferred until after the `load` event, this is too late to handle messages sent from the service worker for navigation events, since [`postMessage()` buffering only occurs prior to the DOMContentLoaded event](https://html.spec.whatwg.org/#the-end).

One solution to deal with this would be to add the listener in the constructor, but then that would increase the chances that messages are received from unexpected service workers (as discussed in #2201).

Whatever solution we choose for this issue should take into account the desired behavior for #2201 as well.
